--- a/lib/vssh/ssh.h
+++ b/lib/vssh/ssh.h
@@ -143,7 +143,7 @@
   const char *authlist;       /* List of auth. methods, managed by libssh2 */
 
   /* common */
-  const char *passphrase;     /* pass-phrase to use */
+  char *passphrase;           /* strdup'ed pass-phrase to use or NULL */
   char *pub_key;              /* strdup'ed public key file */
   char *priv_key;             /* strdup'ed private key file */
   sshstate state;             /* always use ssh.c:state() to change state! */
--- a/lib/vssh/vssh.c
+++ b/lib/vssh/vssh.c
@@ -427,9 +427,14 @@
         goto fail;
     }
 
-    sshc->passphrase = data->set.ssl.primary.key_passwd;
-    if(!sshc->passphrase)
-      sshc->passphrase = "";
+    {
+      const char *keypasswd = CURL_EASY_STR(data, STRING_KEY_PASSWD);
+      if(keypasswd) {
+        sshc->passphrase = curlx_strdup(keypasswd);
+        if(!sshc->passphrase)
+          goto fail;
+      }
+    }
 
     if(sshc->pub_key)
       infof(data, "SSH: public key file '%s'", sshc->pub_key);
--- a/lib/vssh/libssh2.c
+++ b/lib/vssh/libssh2.c
@@ -1231,6 +1231,5 @@
   if(rc ||
-     !(attrs.flags & LIBSSH2_SFTP_ATTR_SIZE) ||
-     (attrs.filesize == 0)) {
+     !(attrs.flags & LIBSSH2_SFTP_ATTR_SIZE)) {
     /*
      * libssh2_sftp_open() did not return an error, so maybe the server
      * does not support stat()
@@ -1492,7 +1492,7 @@
                                            user,
                                            curlx_uztoui(strlen(user)),
                                            sshc->pub_key,
-                                           sshc->priv_key, sshc->passphrase);
+                                           sshc->priv_key, sshc->passphrase ? sshc->passphrase : "");
   if(rc == LIBSSH2_ERROR_EAGAIN)
     return CURLE_AGAIN;
 
@@ -2555,6 +2555,7 @@
   DEBUGASSERT(!sshc->kh);
   DEBUGASSERT(!sshc->ssh_agent);
 
+  curlx_safefree(sshc->passphrase);
   curlx_safefree(sshc->pub_key);
   curlx_safefree(sshc->priv_key);
   curlx_safefree(sshc->quote_path1);
